src/core/service.c

```diff
@@ -2865,7 +2864,9 @@ static void service_notify_cgroup_empty_event(Unit *u) {
                  * SIGCHLD for. */
 
         case SERVICE_START:
-                if (s->type == SERVICE_NOTIFY) {
+                if (s->type == SERVICE_NOTIFY &&
+                    main_pid_good(s) == 0 &&
+                    control_pid_good(s) == 0) {
```

Comment:
This still doesn't take care of
```
[Service]
Type=notify
NotifyAccess=all
RemainAfterExit=yes
ExecStart=/bin/bash -c '(sleep 3; /usr/bin/systemd-notify --ready) &'
```
which works without PAMName= and doesn't with.

Reply:
as mentioned above, i am pretty sure this should be considered expected behaviour. It does deserve some documentation however. Will add.

Reply:
As far as I can tell, no control processes can be active when we are in a SERVICE_START state (a reload control process can be started, but immediately changes the state). Thus we can drop `&& control_pid_good(s)`.

I thought about a `notify_start_good` function that determines whether successful startup is still possible. This could then be used here and at [these lines](https://github.com/systemd/systemd/blob/v234/src/core/service.c#L2924-L2925) with the benefit of having a single place where the non-trivial logic would be implemented. However, it may be easier to leave the current check at `s->type == SERVICE_NOTIFY && main_pid_good(s) == 0` and change [these lines](https://github.com/systemd/systemd/blob/v234/src/core/service.c#L2924-L2925) into
```
else if (!s->remain_after_exit || s->notify_access == NOTIFY_MAIN)
        /* The service has never been and will never be active */
```

~~----~~

Regarding
> in that case you did the most you can do, to not be associated with the original unit

I feel less strongly so and although I agree that it is better to first handoff 'main' to the child process, I still think it may be more sensible to turn `unit_attach_pids_to_cgroup` into a proper cgroup migration, watch and all.